src/rules/LinearAlgebra/blas.jl

```diff
@@ -0,0 +1,128 @@
+    @eval begin
+        function EnzymeRules.forward(
+            func::Const{typeof(BLAS.$fname)},
+            RT::Type{<:Union{Const,DuplicatedNoNeed,Duplicated}},
+            n::Const{<:Integer},
+            X::ConstOrDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incx::Const{<:Integer},
+            Y::ConstOrDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incy::Const{<:Integer},
+        ) where {T<:BLAS.$Ttype}
+            dval = if !(X isa Const) && !(Y isa Const)
+                func.val(n.val, X.dval, incx.val, Y.val, incy.val) +
+                func.val(n.val, X.val, incx.val, Y.dval, incy.val)
+            elseif !(X isa Const)
+                func.val(n.val, X.dval, incx.val, Y.val, incy.val)
+            elseif !(Y isa Const)
+                func.val(n.val, X.val, incx.val, Y.dval, incy.val)
+            else
+                nothing
+            end
+
+            if RT <: Const
+                return nothing
+            elseif RT <: DuplicatedNoNeed
+                dval === nothing ? zero(T) : dval
+            else
+                val = func.val(n.val, X.val, incx.val, Y.val, incy.val)
+                return Duplicated(val, dval === nothing ? zero(T) : dval)
+            end
+        end
+
+        function EnzymeRules.augmented_primal(
+            config::EnzymeRules.ConfigWidth{1},
+            func::Const{typeof(BLAS.$fname)},
+            RT::Type{<:Union{Const,Active,Duplicated}},
+            n::Const{<:Integer},
+            X::ConstOrDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incx::Const{<:Integer},
+            Y::ConstOrDuplicated{<:Union{Ptr{T},AbstractArray{T}}},
+            incy::Const{<:Integer},
+        ) where {T<:BLAS.$Ttype}
+            primal, shadow = _maybe_primal_shadow(
+                config, func.val, (n.val, X.val, incx.val, Y.val, incy.val)
+            )
+
+            # build tape
+            if !(RT <: Const)
+                _, _, Xow, _, Yow = EnzymeRules.overwritten(config)
+                if Xow || BLAS.$fname === BLAS.dotu
+                    Xtape = _strided_tape(n.val, X.val, incx.val)
+                else
+                    Xtape = nothing
+                end
+                if Yow || BLAS.$fname === BLAS.dotu
+                    Ytape = _strided_tape(n.val, Y.val, incy.val)
+                else
+                    Ytape = nothing
+                end
+                if BLAS.$fname === BLAS.dotu
+                    conj!(Xtape)
+                    conj!(Ytape)
+                end
+                tape = (Xtape, Ytape)
+            else
+                tape = nothing
+            end
+
+            return EnzymeRules.AugmentedReturn(primal, shadow, tape)
+        end
+
+        function EnzymeRules.reverse(
+            config::EnzymeRules.ConfigWidth{1},
+            ::Const{typeof(BLAS.$fname)},
+            dret::Union{Active,Type{<:Const}},
```

Comment:
@wsmoses the rules for `dot` and `dotc` are defined identically (the former for real inputs, the latter for complex inputs), but for `dotu` Enzyme wants to provide `dret` as a `Type{<:Duplicated}`, which wouldn't contain derivative information (i.e. `.dval`). Why is this?

```julia
julia> using Enzyme, LinearAlgebra

julia> n = 10;

julia> x, y = randn(n), randn(n);

julia> ∂x, ∂y = map(zero, (x, y));

julia> autodiff(Reverse, BLAS.dot, Active, Const(n), Duplicated(x, ∂x), Const(1), Duplicated(y, ∂y), Const(1))
┌ Warning: Using fallback BLAS replacements, performance may be degraded
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/BxfIW/src/utils.jl:56
((nothing, nothing, nothing, nothing, nothing),)

julia> x, y = randn(ComplexF64, n), randn(ComplexF64, n);

julia> ∂x, ∂y = map(zero, (x, y));

julia> autodiff(Reverse, BLAS.dotc, Active, Const(n), Duplicated(x, ∂x), Const(1), Duplicated(y, ∂y), Const(1))
┌ Warning: Using fallback BLAS replacements, performance may be degraded
└ @ Enzyme.Compiler ~/.julia/packages/GPUCompiler/BxfIW/src/utils.jl:56
ERROR: Enzyme execution failed.
Enzyme: No custom reverse rule was appliable for Tuple{EnzymeCore.EnzymeRules.ConfigWidth{1, true, true, (false, false, false, false, false, false)}, Const{typeof(LinearAlgebra.BLAS.dotc)}, Type{Duplicated{ComplexF64}}, Tuple{Nothing, Nothing}, Const{Int64}, Duplicated{Vector{ComplexF64}}, Const{Int64}, Duplicated{Vector{ComplexF64}}, Const{Int64}}
```